lib/DefinePlugin.js

```diff
@@ -52,15 +64,26 @@ class DefinePlugin {
 					const walkDefinitions = (definitions, prefix) => {
 						Object.keys(definitions).forEach(key => {
 							const code = definitions[key];
+
 							if (
 								code &&
 								typeof code === "object" &&
+								!(code instanceof RuntimeValue) &&
 								!(code instanceof RegExp)
 							) {
 								walkDefinitions(code, prefix + key + ".");
 								applyObjectDefine(prefix + key, code);
 								return;
 							}
+
+							if (code instanceof RuntimeValue && parser.state) {
+								for (const fileDependency of code.fileDependencies) {
+									parser.state.module.buildInfo.fileDependencies.add(
+										fileDependency
+									);
+								}
```

Comment:
This would add the dependencies to every module, but it should only be added to modules where the define is used. You can move this code into the `expression`, `evaluate`, `evaluateTypeof` hooks to be only executed when define is used. You should also only execute the passed function when one of these hooks is triggered.

Reply:
Maybe you need to move `code = toCode(code)` into the hook function itself.

btw.

``` js
parser.hooks.expression.for(key).tap("...", ParserHelpers.toConstantDependency(parser, code))
// is equal to
parser.hooks.expression.for(key).tap("...", expr => {
  return ParserHelpers.toConstantDependency(parser, code));
})
```

You probably need to rewrite this to add your own code.